src/smtp_observer.py

```diff
@@ -0,0 +1,147 @@
+            user = relation_data.user
+            password = self._get_password_from_relation_data(relation_data)
+
+        # Not all combinations for the next variables are correct. See:
+        # https://github.com/matrix-org/synapse/blob/develop/synapse/config/emailconfig.py
+        force_tls = False
+        enable_tls = False
+        require_transport_security = False
+        if relation_data.transport_security == TransportSecurity.STARTTLS:
+            enable_tls = True
+            require_transport_security = True
+        elif relation_data.transport_security == TransportSecurity.TLS:
+            force_tls = True
+            enable_tls = True
+            require_transport_security = True
+
+        return SMTPConfiguration(
+            enable_tls=enable_tls,
+            force_tls=force_tls,
+            require_transport_security=require_transport_security,
+            host=relation_data.host,
+            port=relation_data.port,
+            user=user,
+            password=password,
+        )
+
+    def _get_password_from_relation_data(self, relation_data: SmtpRelationData) -> Optional[str]:
+        """Get smtp password from relation data.
+
+        Arguments:
+            relation_data: The relation data from where to extract the password
+
+        Returns:
+            the password or None if no password found
+        """
+        # If the relation data password_id exists, that means that
+        # Juju version is >= 3.0 and secrets are used for the password.
+        # Otherwise, use the field password as a fallback
+        if relation_data.password_id:
+            secret = self.model.get_secret(id=relation_data.password_id)
+            content = secret.get_content()
+            return content["password"]
+        return relation_data.password
+
+    def _on_smtp_relation_data_available(self, _: SmtpDataAvailableEvent) -> None:
+        """Handle SMTP data available."""
+        self.model.unit.status = ops.MaintenanceStatus("Preparing the SMTP integration")
+        logger.debug("_on_smtp_data_available: Enabling SMTP")
+        self._enable_smtp()
+
+    def _enable_smtp(self) -> None:
+        """Enable  SMTP."""
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect() or self._pebble_service is None:
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
+        try:
+            self._pebble_service.enable_smtp(container)
+        except PebbleServiceError as exc:
+            self._charm.model.unit.status = ops.BlockedStatus(f"SMTP integration failed: {exc}")
+            return
+        self._charm.unit.status = ops.ActiveStatus()
+
+    @property
+    def _pebble_service(self) -> typing.Any:
+        """Return instance of pebble service.
+
+        Returns:
+            instance of pebble service or none.
+        """
+        return getattr(self._charm, "pebble_service", None)
```

Comment:
I think it might help to have this be injected if circular dependency is not a problem.

Reply:
As it is now, observers are created first, from them the state is built and then the pebble service gets the state injected. I believe this should be better refactored in another PR, changing all the observers.

See https://github.com/canonical/synapse-operator/blob/34c5118a4fc8a070cb6d43a82436e82348fea605/src/charm.py#L51-L64